### ISSUE TITLE
Binder does not apply appropriate special case for unbound local variables used within class body

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3454]
See this issue in mypy's issue tracker: https://github.com/python/mypy/issues/12764. Pyright has the same bug.

Here's a simplified sample.

```python
from typing import TYPE_CHECKING

x = 0

def f() -> None:
    x = "a"

    class C:
        print(x)  # Prints 0, not "a" as you might expect
        if TYPE_CHECKING:
            reveal_type(x)  # Should be int, not str
        x = "b"
        print(x)  # Prints "b" as expected

f()
```

Note that if you add the statement `nonlocal x` at the top of the class definition, it will print `"a"` instead.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.247, which I just published. It will also be included in the next release of pylance.
